Python3 support

thrix opened the pull request:
Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
Python3 support

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

happz commented:
Well, since you kicked this feature off, few notes from my checklist:

* The very first point must be to enable Python 3 in our `tox.ini`, changing `py27` to `{py27,py35,py36,...}` or whatever other versions we'd like to support. That should get us unit tests and static analysis in Python 3 environment.
* Use `2to3` to find out what it would do to update our sources, but don't let it apply suggested changes. it should discover stuff like `iterkeys` => `keys`, `print` vs `print()` and so on, but we should use `six` to fix those issues. `2to3` provides nice starting point, though.
* `six` is really awesome, and the good approach is to open its documentation and grep our sources for every feature it provides. E.g. we're using `isinstance(foo, int)`, this should be changed to `isinstance, foo, six.integer_types)`. The same applies to renamed and moved functions and objects, every time we use such, we should simply use the one from `six`.
* Strings :/ This will be quite painful, I'm afraid. We must check for every `str` or `unicode` we use, and follow the trail of calls and find out boundaries on which we should decide what data type the boundary accepts - whether (unicode) string, or bytes, and convert where necessary. Python 3 uses Unicode strings by default, while we're quite limited to pure ASCII strings, and this will be even worse when we take the "Python 3 ready" `gluetool`, and use it with our custom modules - then we're going to run into a shitstorm, like passing Unicode to commands that cannot deal with that, which would force us to apply other conversions, and so on.

It's doable, and it *is* relatively easy task, yet it's still a lot of work :/ Unfortunately there's no such thing as "partial Python 3 support", that's even worse than "no Python 3 support at all", and until everything is fixed (with many more test cases to feed our functions things like Unicode strings) it cannot be merged because that'd complicate things of every user of using `gluetool` by installing lot of exceptions and unexpected behaviors :(
I'd be happy to help if this feature is of such a high priority that it made you began to work on it :)

thrix committed:
add py36 to tox.ini

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix commented:
@happz thanks for the info, very valuable. It is not a big priority. Let's just slowly move toward it :),

optak commented on gluetool/utils.py:
```diff
@@ -14,8 +14,7 @@
-import urllib2
-
+from six.moves import urllib
 import bs4
```
very nice

thrix replied:
yeah, nice

thrix committed:
Use urlnormalizer instead of urlnorm

The latter modules has no propper Python3 support yet.

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix commented:
All 2to3 issues resolved (1st try), one thing remains, which feels like a bug in 2to3:

```
RefactoringTool: Skipping optional fixer: buffer
RefactoringTool: Skipping optional fixer: idioms
RefactoringTool: Skipping optional fixer: set_literal
RefactoringTool: Skipping optional fixer: ws_comma
RefactoringTool: No changes to gluetool/__init__.py
RefactoringTool: No changes to gluetool/color.py
RefactoringTool: No changes to gluetool/glue.py
RefactoringTool: No changes to gluetool/help.py
RefactoringTool: No changes to gluetool/log.py
RefactoringTool: No changes to gluetool/proxy.py
RefactoringTool: No changes to gluetool/sentry.py
RefactoringTool: Can't parse gluetool/tool.py: ParseError: bad input: type=22, value='=', context=('', (214, 22))
RefactoringTool: No changes to gluetool/utils.py
RefactoringTool: No changes to gluetool/version.py
RefactoringTool: No changes to gluetool/pylint/__init__.py
RefactoringTool: No changes to gluetool/pylint/shared_defined.py
RefactoringTool: Can't parse gluetool/pylint/unknown_option.py: ParseError: bad input: type=4, value='\n', context=('', (71, 52))
RefactoringTool: No changes to gluetool/tests/__init__.py
RefactoringTool: No changes to gluetool/tests/conftest.py
RefactoringTool: No changes to gluetool/tests/test_core.py
RefactoringTool: No changes to gluetool/tests/test_error.py
RefactoringTool: Refactored gluetool/tests/test_json.py
RefactoringTool: No changes to gluetool/tests/test_load_yaml.py
RefactoringTool: No changes to gluetool/tests/test_log_exception.py
RefactoringTool: No changes to gluetool/tests/test_logging.py
RefactoringTool: No changes to gluetool/tests/test_new_xml_element.py
RefactoringTool: No changes to gluetool/tests/test_normalize_option.py
RefactoringTool: No changes to gluetool/tests/test_pipeline_step.py
RefactoringTool: No changes to gluetool/tests/test_render_template.py
RefactoringTool: No changes to gluetool/tests/test_run_command.py
RefactoringTool: No changes to gluetool/tests/test_shared.py
RefactoringTool: No changes to gluetool/tests/test_treat_url.py
RefactoringTool: No changes to gluetool/tests/test_wait.py
RefactoringTool: Files that need to be modified:
RefactoringTool: gluetool/__init__.py
RefactoringTool: gluetool/color.py
RefactoringTool: gluetool/glue.py
RefactoringTool: gluetool/help.py
RefactoringTool: gluetool/log.py
RefactoringTool: gluetool/proxy.py
RefactoringTool: gluetool/sentry.py
RefactoringTool: gluetool/utils.py
RefactoringTool: gluetool/version.py
RefactoringTool: gluetool/pylint/__init__.py
RefactoringTool: gluetool/pylint/shared_defined.py
RefactoringTool: gluetool/tests/__init__.py
RefactoringTool: gluetool/tests/conftest.py
RefactoringTool: gluetool/tests/test_core.py
RefactoringTool: gluetool/tests/test_error.py
RefactoringTool: gluetool/tests/test_json.py
RefactoringTool: gluetool/tests/test_load_yaml.py
RefactoringTool: gluetool/tests/test_log_exception.py
RefactoringTool: gluetool/tests/test_logging.py
RefactoringTool: gluetool/tests/test_new_xml_element.py
RefactoringTool: gluetool/tests/test_normalize_option.py
RefactoringTool: gluetool/tests/test_pipeline_step.py
RefactoringTool: gluetool/tests/test_render_template.py
RefactoringTool: gluetool/tests/test_run_command.py
RefactoringTool: gluetool/tests/test_shared.py
RefactoringTool: gluetool/tests/test_treat_url.py
RefactoringTool: gluetool/tests/test_wait.py
RefactoringTool: There were 2 errors:
RefactoringTool: Can't parse gluetool/tool.py: ParseError: bad input: type=22, value='=', context=('', (214, 22))
RefactoringTool: Can't parse gluetool/pylint/unknown_option.py: ParseError: bad input: type=4, value='\n', context=('', (71, 52))
--- gluetool/tests/test_json.py	(original)
+++ gluetool/tests/test_json.py	(refactored)
@@ -1,7 +1,7 @@
 import json
 import re
 import string
-from __future__ import print_function
+
 
 import pytest
 from hypothesis import assume, given, strategies as st
@@ -53,4 +53,4 @@
 
     with pytest.raises(gluetool.GlueError,
                        match=r"(?ms)Unable to load JSON file '{}': Expecting object: line 1 column 1 \(char 0\)".format(re.escape(filepath))):
-        print(load_json(filepath))
+        print((load_json(filepath)))
```

thrix committed:
fix all 2to3 issues identified

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

happz commented:
@thrix Could you please rebase the patch? It's kind of messy, mixed with patches already merged into master.

thrix commented:
@happz right, rebased ... still needs work .. I am bit concerned about my fixes to unicode/str as I see now (wtf the jinja2 traceback):

```
$ gluetool -h

!!! While handling an exception, another one appeared !!!

Will try to submit it to Sentry but giving up on everything else.

!!! While submitting an exception to the Sentry, another exception appeared !!!
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>     Giving up on everything...
<_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>
Traceback (most recent call last):
  File "/home/mvadkert/git/github.com/gluetool/gluetool/tool.py", line 39, in wrapped
    return func(self, *args, **kwargs)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/tool.py", line 272, in setup
    Glue = self.Glue = gluetool.Glue(tool=self, sentry=self.sentry)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 1326, in __init__
    super(Glue, self).__init__()
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 404, in __init__
    Configurable._for_each_option_group(_verify_options, self.options)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 364, in _for_each_option_group
    callback(group_options, group_name=group_name)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 402, in _verify_options
    Configurable._for_each_option(_verify_option, options)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 341, in _for_each_option
    callback(name, names, params)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/glue.py", line 397, in _verify_option
    params['help'] = option_help(params['help'])
  File "/home/mvadkert/git/github.com/gluetool/gluetool/help.py", line 264, in option_help
    return ' '.join(processed.splitlines())
TypeError: sequence item 0: expected str instance, bytes found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mvadkert/git/github.com/gluetool/gluetool/tool.py", line 205, in _handle_failure
    self._handle_failure_core(failure)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/tool.py", line 192, in _handle_failure_core
    msg = "Exception raised: {}".format(failure.exc_info[1].message)
AttributeError: 'TypeError' object has no attribute 'message'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mvadkert/git/github.com/gluetool/gluetool/tool.py", line 221, in _handle_failure
    print(gluetool.log.LoggingFormatter._format_exception_chain(sys.exc_info()), file=sys.stderr)
  File "/home/mvadkert/git/github.com/gluetool/gluetool/log.py", line 411, in _format_exception_chain
    _add_block('Exception', exc_info[1], exc_info[2])
  File "/home/mvadkert/git/github.com/gluetool/gluetool/log.py", line 407, in _add_block
    output.append(tmpl.render(label=label, exception=exc, stack=stack))
  File "/home/mvadkert/venvs/gluetool3/lib/python3.6/site-packages/Jinja2-2.9.6-py3.6.egg/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/mvadkert/venvs/gluetool3/lib/python3.6/site-packages/Jinja2-2.9.6-py3.6.egg/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/mvadkert/venvs/gluetool3/lib/python3.6/site-packages/Jinja2-2.9.6-py3.6.egg/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/mvadkert/venvs/gluetool3/lib/python3.6/site-packages/Jinja2-2.9.6-py3.6.egg/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 11, in top-level template code
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'iterkeys'
```

thrix committed:
Fix additional Py3 issues

1. iterkeys is gone from dict in Py3, use items() which is fine in this
case, as we do not expect the f_locals dict to be very huge

2. Typerror.message is not around anymore, use directly the string
conversion instead

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
Additional Python3 fixes

1. Do not use depreprecated message attribute of Exception

2. docutils.core.publish_string returns bytes, convert it to string

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix commented:
@happz so we have a first "working" version, at least with no traceback when I try

```
$ gluetool -dl
[09:41:24] [D] logger set up: output_file='None', level=10
[09:41:24] [D] loading modules from these paths:
[
    "/home/mvadkert/venvs/gluetool3/gluetool_modules"
]
[09:41:24] [D] pipeline description:
[]
Available modules

  -- no modules found --
[09:41:24] [D] Exiting with status 0
```

You can start slight reviewing if you have time.

thrix committed:
Fix additional iteritems

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix commented:
still working on tests ...

thrix commented:
hmm, python2 is now broken, also needs love

happz commented on gluetool/glue.py:
```diff
@@ -1,14 +1,18 @@
+try:
+    import configparser
+except:
+    from six.moves import configparser
```
Just `from six.moves import configparser`, no need to bother with import error, `six` will take care of it.
Also, group it with other imports from `six`.

happz commented on gluetool/glue.py:
```diff
@@ -327,7 +331,7 @@ def _for_each_option(callback, options):
         # Sort options by their names - no code has a strong option on their order, so force
         # one to all users of this helper.
-        option_names = sorted(options.keys(), key=lambda x: x[1] if isinstance(x, tuple) else x)
+        option_names = sorted(list(options.keys()), key=lambda x: x[1] if isinstance(x, tuple) else x)
```
This should be `sorted(iterkeys(options))` - `sorted` should get an iterator, it's not necessary to materialize list, `sorted` will handle things internally much better.

thrix replied:
I see, will be fixed in next patches, thanks

happz commented on gluetool/help.py:
```diff
@@ -161,7 +164,13 @@ def rst_to_text(text):
+    text = docutils.core.publish_string(text, writer=sphinx.writers.text.TextWriter(DummyTextBuilder))
+
+    # in Python3 the publish_string returns bytes, convert it to string
+    if isinstance(text, bytes):
```
~I'd rather check its callers, I see no reason why `rst_to_text` should be called with bytes in Python 3. It's used to process strings, mostly docstrings, those should be `str` :/~

Nevermind, `docutils` are giving us `bytes`? This might be also worth checking but I guess it's less priority. It's good that `rst_to_text` returns always a string.

thrix replied:
yes, seems so:
```
In [9]: type(docutils.core.publish_string("test"))
Out[9]: bytes
```

happz commented on gluetool/log.py:
```diff
@@ -113,7 +113,7 @@ def warn_sentry(self, message, *args, **kwargs):
     {{ text | default('') }}
 
-    Local variables:{% for name in frame.f_locals.iterkeys() | sort %}
+    Local variables:{% for name in frame.f_locals.items() | sort %}
```
I'm kind of worried about this, as `iteritems` and `items` differ in Pythons 2 and 3, and `items` will materialize list of items when being run under Python 2. We might use something like this, maybe in a simple "workarounds" module or in `utils`, don't know:

```
import six

env = jinja2.Environment(loader=jinja2.FileSystemLoader('your_dir'))
env.globals['six_iteritems'] = six.iteritems
template = env.get_template('your_template')
```

We have similar stuff in `colors` already, to inject our `style` helper into Jinja templates as a filter.

thrix replied:
why are you worried? I guess list of locals can be long, but not so much usually. I do not mind adding a helper, I just do not see the value for this particular case. Am I missing something?

happz replied:
It's just a matter of different behavior under Python2 and Python3. I'm not concerned about the performance issues, as you said, the list of locals won't be long enough to cause any problems. It will just behave differently when running on different Python versions which I'm not a big fan of.

happz commented on gluetool/proxy.py:
```diff
@@ -5,6 +5,8 @@
 # Proxy class comes from the AS recipe
 # http://code.activestate.com/recipes/496741-object-proxying/
 
+from builtins import object
```
Why is this necessary?

thrix replied:
https://stackoverflow.com/questions/36517648/in-python-why-import-object-from-builtins-module)

happz replied:
Ah, I see. Nice.

happz commented on gluetool/proxy.py:
```diff
@@ -33,7 +35,7 @@ def __delattr__(self, name):
         setattr(object.__getattribute__(self, "_obj"), name, value)
 
-    def __nonzero__(self):
+    def __boolo__(self):
```
We need both - Python 2 will try to call ``__nonzero__``.

```python
def __bool__
  return bool(object.__getattribute__(self, "_obj"))

if six.PY2:
  __nonzero__ = __bool__
```

Also, typo: `s/boolo/boo/`

thrix replied:
thanks, will be in next patch

happz commented on gluetool/tests/test_json.py:
```diff
@@ -1,6 +1,7 @@
 import json
 import re
 import string
+from __future__ import print_function
```
I believe `__future__` imports must be the very first import in a file.

thrix replied:
right, thanks, resolved

happz commented on gluetool/utils.py:
```diff
@@ -584,25 +585,20 @@ def treat_url(url, logger=None):
     :param str url: URL to clear.
     :param gluetool.log.ContextAdapter logger: logger to use for logging.
     :rtype: str
+    :raises: gluetool.glue.GlueErrorL: if URL is invalid
```
Typo, `GlueErrorL`

thrix replied:
thanks, fixed

happz commented on gluetool/utils.py:
```diff
@@ -627,7 +623,7 @@ def _render(template):
             return str(template.render(**kwargs).strip())
 
-        if isinstance(template, (str, unicode)):
+        if isinstance(template, str):
```
`six.string_types` instead of `str`

thrix replied:
thanks, fixed

happz commented on gluetool/utils.py:
```diff
@@ -725,7 +721,7 @@ def dump_yaml(data, filepath, logger=None):
 def _json_byteify(data, ignore_dicts=False):
     # if this is a unicode string, return its string representation
-    if isinstance(data, unicode):
+    if isinstance(data, str):
```
`six.string_types` instead of `str`

thrix replied:
thanks, fixed all checks to be python2 and python3 compatible

happz commented on gluetool/utils.py:
```diff
@@ -818,7 +814,7 @@ def __init__(self, filepath, logger=None):
                 raise GlueError("Invalid format: '- <pattern>: <result>' expected, '{}' found".format(pattern_dict))
 
-            pattern = pattern_dict.keys()[0]
+            pattern = list(pattern_dict.keys())[0]
```
Under Python 2, this would a list made of a list. `list(iterkeys(pattern_dict))` should work the same for both.

thrix replied:
thanks, fixed

happz commented on setup.py:
```diff
@@ -56,6 +57,7 @@
               'License :: OSI Approved :: BSD License',
               'Operating System :: POSIX :: Linux',
               'Programming Language :: Python :: 2.7',
+              'Programming Language :: Python :: 3',
```
I'ts better to state the exact version, 3.6 seems to be the current stable.

thrix replied:
thanks fixed

happz commented on tox.ini:
```diff
@@ -13,6 +13,12 @@ commands = pytest -v -ra --cov=gluetool --cov-report=html:coverage-report {posar
+
+[testenv:py36-static-analysis]
+commands = pytest -v -ra --pylint --flake8 -k 'not test_' --ignore=docs/source/module_parsers.py {posargs}
+
```
It might be possible to merge `py27-*` with their `py36-*` counterparts, however I don't know how. But tox hase some abilities, e.g. `envlist = {py27,py36}-{unit-tests,static-analysis}`. I'll find the answer.

happz replied:
Here you go - condensed, and with one exception I really like it. When you start with more than one Python version, tox starts to prove very useful:

``` ini
[tox]
envlist = {py27,py36}-{static-analysis,unit-tests,codecov}

[testenv]
# Tox should set correct Python version given the environment. let's trust it knows how.
#basepython =
#  py27: python2.7
#  py36: python3.6

# Use a single env directory for all environments but separate by Python versions.
#
# It could be easier to just say:
#
#   envdir = {toxworkdir}/{basepython}
#
# but, since it's hardcoded in Tox, basepython is not set when envdir is being examined,
# therefore, the replacement fails, hence the explicit names.
envdir =
  py27: {toxinidir}/.tox/py27
  py36: {toxinidir}/.tox/py36

# Don't spoil our nice virtualenvs with system packages
sitepackages = False

# Just test dependencies - gluetool will pull in its own dependencies when being installed
deps = -rtest-requirements.txt

# Capture coverage per Python version
setenv = COVERAGE_FILE={envdir}/.coverage

# Pass necessary env vars to let CI and coverage gathering play together nicely
passenv = CI TRAVIS TRAVIS_* CODECOV_TOKEN

# Conditionals don't support multiline branches, therefore codecov upload is a separate step.
# but that's fine, thanks to options we give it it knows where to find coverage (specific for
# Python version), and we even tell it to add Python version as a flag - keeping coverage
# separate even in Codecov seems to be a good starting point, we can always start merging
# them later.
#
# And there's an advantage: the codecov upload does not run when one runs *-unit-tests env locally.
# It'd fail anyway, since one does not usually have CODECOV_TOKEN set, and it's still possible
# to run it explicitly to upload one's local coverage in case of need.
commands =
  static-analysis: pytest -v -ra --pylint --flake8 -k 'not test_' {posargs}
  unit-tests: pytest -v -ra --cov=gluetool --cov-report=html:{envdir}/coverage-report {posargs}
  codecov: codecov --file={envdir}/.coverage --flags={basepython}

# [testenv:py27-doctest]
# whitelist_externals = ansible-playbook
# commands = /usr/bin/ansible-playbook generate-docs.yml -e generate_dir="{envtmpdir}/docs"

# [testenv:py27-static-analysis-coala]
# whitelist_externals = docker
# commands = docker run -ti --rm -v {toxinidir}:/gluetool:z --workdir=/gluetool coala/base coala -c /gluetool/.coafile --non-interactive
```

Works like a charm for me, just run `tox` to run all environments. It will run `codecov` upload as well, which would fail as you don't have CODECOV_TOKEN env var set locally, but `tox` will report it as a success - which is perfectly fine, it works perfectly within Travis.

thrix replied:
Thanks @happz, incorporated

happz commented:
This seems to be a worthy document: http://python-future.org/automatic_conversion.html. `futurize` mentioned there has also similar functionality like `six` but I'd rather use `futurize` in a read-only mode first, letting it suggest changes first, before making changes. I've never used it but it seems to be a recommended tool by Python 3 docs (https://docs.python.org/3/howto/pyporting.html), we should at least give it a try and see what it has to say about or code. Not sure how it coexists with `six`, we'll see.

thrix committed:
Fix exceptions

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
dep-list: Python2/3 compatbility fixes

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
Fix few tests, include mock

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
tox.ini: improve ini file for both versions

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
Fixed additional comments from @happz

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

thrix committed:
Additional fixes for static analysis

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

happz commented:
@thrix I'm going to rebase your patch on top of the latest `master` - it should add type annotations, and that should help us dealing with APIs that accept strings vs bytes vs unicode mess.

thrix commented:
@happz sure thing syr!

happz commented:
Few notes, in the sake of sanity:

* Unicode strings everywhere, both Python 2 & 3. Any `bytes` or `str` input must be decoded to `unicode` when entering gluetool from external source, and encoded back to `bytes`/`str` when leaving - gluetool's API talks Unicode as much as possible;
* bumped version of `six` - `ensure_{text,str,binary}` are extremely useful to transparently encode/decode strings (e.g. when calling Python 2's standard library functions that accept `str` and our function was given `unicode`), and they reduce nicely to no-ops where the action makes no sense (e.g. `ensure_text(unicode)` is no-op);
* bumped versions of `mypy` & co. - not good enough yet, we need `mypy` with `typeshed` that includes stubs for `six.ensure_*` (merged & unreleased yet: https://github.com/python/typeshed/pull/2704/files)

`2to3` is reasonably quiet, the same goes for `futurize`. Next step: verify inputs from external world (user input (command-line options, ...), files (config, yaml, ...), network (fetch URL, ...) convert strings and bytes correctly, and make unit tests pass.

BTW I was able to load `execute-command` module, but it didn't work correctly, e.g. all logs were dumped into a debug file, `stderr` remained empty. It's probably correct, but some APIs might have changed in Python 3, and we're hooking to some not really standard places when it comes to logging. But it's going to be fine.

PS: force-push due to the rebase.

happz committed:
Rebased to master for type annotations

- rebase brought type annotations
- gluetool's methods accept unicode only - lingua franca
- used six to make conversions with less rough edges

Passed mypy tests

happz committed:
Making test suite happy - static analysis, unit tests & type check passed

happz committed:
Remove non-matrix Python versions

happz commented:
Type check test fails due to lack of `ensure_*` stubs for mypy, hopefully new `typeshed` would be released soon. Works for me when I patch `typeshed` manually.

happz commented:
So, what's left?

- [ ] re-check output of `2to3`
- [ ] re-check output of `futurize`
- [ ] decide what type to accept as exception messages - `raise Foo('abc')` reads simpler than `raise Foo(u'abc')` but forcing Unicode even here makes sense, exceptions are not the outer border after all
- [ ] ~the same goes for logging api - what should be the type of string given to `self.info`? `Text` or `AnyStr` & `ensure_text` call inside? Adding `u` to all strings seems like way too much work but it may save us from insanity later when hunting issues with mixing different types in Python 3~ Probably no longer relevant after LoggerMixin leveled the playfield.
- [ ] check logging API & make sure it handles Unicode correctly when handing records back to standard lib - the border might be different in Python 2 & 3
- [ ] document uses of `ensure_*` - when they are used, they are obviously converting some sort of input, therefore good comment or docs on *why* the `ensure_*` was used should lie nearby
- [ x] make sure new `mypy` is released with stubs for `six.ensure_*`
- [ ] check with common modules (gluetool bundled, `rules-engine`, `execute-command`, ...), trying to run them in both Python environments, including their tests
- [ ] peer review :)

happz committed:
Export Six' iter* to templates

This makes iterating over mappings in templates compatible with both
Python 2 and 3.

happz committed:
Use next to get the first item of an iterator

Instead of materializing the whole list & using an index.

happz committed:
Gluetool's bundled modules are now Python 3 compatible

happz committed:
Fix bash-completion module, somehow it missed the last patch

happz committed:
Catching up with LoggerMixin and other changes - huge rebase

happz committed:
Fix export of six iter* to Jinja

happz commented:
```
(gluetool3.6) [pts-3:7]: mprchlik@multivac [python3] ~/virtualenvs/gluetool3.6/repo $ gluetool -l --module-path gluetool_modules/
Available modules

    bash-completion                  Generate Bash completion configuration.
    dep-list                         Collect module dependencies for citool modules
    yaml-pipeline                    Runs pipeline, described by a given YAML file.
(gluetool3.6) [pts-3:7]: mprchlik@multivac [python3] ~/virtualenvs/gluetool3.6/repo $ 
```
And tests are passing with both Python 2.7 and Python 3.6.

So, what's left?

- [x] re-check output of `2to3` - it may report things we don't want to fix (e.g. `'foo' => u'foo') because it doesn't care about Python2 compatibility!
- [x] re-check output of `futurize` - it may report things we don't want to fix (e.g. `'foo' => u'foo') because it doesn't care about Python2 compatibility!
- [ ] document uses of `ensure_*` - when they are used, they are obviously converting some sort of input, therefore good comment or docs on *why* the `ensure_*` was used should lie nearby
- [x] make sure new `mypy` is released with stubs for `six.ensure_*`
- [ ] check with common modules (gluetool bundled, `rules-engine`, `execute-command`, ...), trying to run them in both Python environments, including their tests
- [ ] peer review :)

happz committed:
Remove spurious print()

happz committed:
Getting rid of (obsoleted) Text

happz committed:
Fix use of jinja in help

ukulekek commented on gluetool/glue.py:
```diff
@@ -24,14 +26,15 @@
 
 # Type annotations
-# pylint: disable=unused-import,wrong-import-order
+# pylint: disable=unused-import,wrong-import-order,
```
excess comma

happz replied:
Ack, fixed.

ukulekek commented on gluetool/help.py:
```diff
@@ -88,6 +107,7 @@ def C_LITERAL(text):
 _original_TextTranslator = sphinx.writers.text.TextTranslator
 
 
+# ##pylint: disable=abstract-method
```
excess hashes

happz replied:
Ack, fixed.

ukulekek commented on gluetool/help.py:
```diff
@@ -206,12 +227,12 @@ def rst_to_text(text):
 
-    :param str text: string to render.
-    :rtype: str
+    :param text text: string to render.
```
What is `text` type?

happz replied:
It was an attempt to cover all variants of "string" in both Python 2 and 3, but I changed it back to `str` for now.

ukulekek commented on gluetool/utils.py:
```diff
@@ -522,8 +529,8 @@ def _communicate_inspect(self, inspect_callback):
         if inspect_callback is None:
-            def stdout_write(stream, data, flush=False):
-                # type: (Any, Any, bool) -> None
+            def stdout_write(stream, data, flush):
```
Is `flush` important here? It's unused in the method

ukulekek replied:
I see. Never mind

mkluson commented on gluetool/tests/test_json.py:
```diff
@@ -50,6 +54,11 @@ def test_load(json_data, log, tmpdir):
-    with pytest.raises(gluetool.GlueError,
-                       match=r"(?ms)Unable to load JSON file '{}': Expecting object: line 1 column 1 \(char 0\)".format(re.escape(filepath))):
-        print load_json(filepath)
+    if six.PY3:
```
`if six.PY2:` is used in the rest of code, could you make it consistent?

happz replied:
Right, will be fixed in the next commit.

happz committed:
String & slot tweaks

happz committed:
Typo fix

mkluson commented on gluetool/tool.py:
```diff
@@ -256,10 +257,11 @@ def _handle_failure(self, failure):
+                print(file=sys.stderr)
+                # pylint: disable=line-too-long
+                print('!!! While submitting an exception to the Sentry, another exception appeared !!!', file=sys.stderr)
+                print(sys.stderr, '    Giving up on everything...', file=sys.stderr)
```
Why do you use `sys.stderr` as a first parameter?

happz replied:
Forgot to get rid of it. Fixed.

mkluson commented on gluetool/tool.py:
```diff
@@ -256,10 +257,11 @@ def _handle_failure(self, failure):
-                print >> sys.stderr, '!!! While logging an exception, another exception appeared !!!'
-                print >> sys.stderr, '    Giving up on everything...'
-                print >> sys.stderr
+                print(file=sys.stderr)
```
What about to add `\n` to the beginning of next print?

happz replied:
I replaced these with multiline strings - it is more readable in Python 2, with multiple `print` statements, in Python 3 it sucks.

mkluson commented on gluetool_modules/bash_completion.py:
```diff
@@ -87,7 +91,7 @@ def _add_options_from_group(options, **kwargs):
 
         # Inspect all option groups defined by the module, and add every option found
-        for module_name in self.glue.modules.iterkeys():
+        for module_name in iterkeys(self.glue.modules):
```
Note: also this change is required https://github.com/gluetool/gluetool/pull/81/files

happz replied:
I believe this one will get in "for free" - either we merge your PR first, then the change would get into this patch via rebase, or we'd merge this PR first and then your PR would get into `master` anyway.

happz committed:
Addressing mkluson's comments

happz committed:
Addressing efedin's comments

happz committed:
ensure_str moved into treat_url

thrix commented:
I guess we can now remove the WIP keyword

thrix reviewed (commented):
I did not find anything, I tried :(

thrix commented:
I am fine with merging this, thanks @happz for all the hard work!!

@ukulekek @optak @mkluson @happz .. please add review, I cannot review, as I am the creator of the original PR.

happz commented:
I'm going to add comments to (hopefully all) `ensure_*` calls and why they are used, that's the last bit left on my TODO I can influence.

akhait reviewed (dismissed):
LGTM

optak reviewed (dismissed):
looks ok, corner cases can be caught later in staging sentry

happz committed:
Tweak of str issues after running real pipelines

happz commented:
Thanks for all reviews, I consider this done (for this moment, until the first bug appears :)
